connect-inject/health_check_handler.go

```diff
@@ -0,0 +1,317 @@
+package connectinject
+
+// handler skeleton inspired by https://github.com/trstringer/k8s-controller-core-resource
+
+import (
+	ctx "context"
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/cenkalti/backoff"
+	"github.com/hashicorp/consul-k8s/subcommand/flags"
+	"github.com/hashicorp/consul/api"
+	log "github.com/hashicorp/go-hclog"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// HCHandler interface contains the methods that are required
+type HCHandler interface {
+	Init() error
+	ObjectCreated(obj interface{}) error
+	ObjectDeleted(obj interface{}) error
+	ObjectUpdated(objNew interface{}) error
+	Reconcile() error
+}
+
+const (
+	healthCheckPassing  = "passing"
+	healthCheckCritical = "critical"
+)
+
+// HealthCheckHandler is a sample implementation of HCHandler
+type HealthCheckHandler struct {
+	Log                log.Logger
+	AclConfig          api.NamespaceACLConfig
+	Client             *api.Client
+	HFlags             *flags.HTTPFlags
+	Clientset          kubernetes.Interface
+	ConsulClientScheme string
+	ConsulPort         string
+}
+
+// getConsulHealthCheckID deterministically generates a health check ID that will be unique to the Agent
+// where the health check is registered and deregistered.
+func (t *HealthCheckHandler) getConsulHealthCheckID(pod *corev1.Pod) string {
+	return fmt.Sprintf("%s_%s_kubernetes-health-check-ttl", pod.Namespace, t.getConsulServiceID(pod))
+}
+
+// return the serviceID of the connect service
+func (t *HealthCheckHandler) getConsulServiceID(pod *corev1.Pod) string {
+	return pod.Name + "-" + pod.Annotations[annotationService]
+	// TODO: I'd like to have this put into an annotation by this point, unfortunately it cannot
+	// be injected by the webhook because pod.Name isnt determined by then
+	//return pod.Annotations[annotationServiceID]
+}
+
+// updateConsulClient updates the Consul Client metadata to point to a new hostIP:port
+// which is the IP of the host that the Pod runs on, in order to make Agent calls locally
+// for health check registration/deregistration.
+func (t *HealthCheckHandler) updateConsulClient(pod *corev1.Pod) error {
+	var err error
+	// We use the whole schema including http/https here because that is normally set in the
+	// CONSUL_HTTP_ADDR environment variable, without https the client will default to http even if the port
+	// is 8501
+	httpFmt := "http"
+	if t.ConsulPort == "8501" {
+		httpFmt = "https"
+	}
+	newAddr := fmt.Sprintf("%s://%s:%s", httpFmt, pod.Status.HostIP, t.ConsulPort)
+	t.HFlags.SetAddress(newAddr)
+
+	// Set client api to point to the new host IP
+	// TODO: APIClient() reads the ENV, see if we can avoid this for perf
+	t.Client, err = t.HFlags.APIClient()
```

Comment:
Could we have this return a new client instead of using a shared `t.Client` state everywhere?